File: backend/src/api/plan/plan.py
"""A Complete User Plan."""


class Plan:
    name: str
    # maps (year, sem) to list of chosen courses
    course_dates: dict[tuple[int, int], list[str]]
    course_reqs: dict[str, list[str]]
    courses: list[str]
    degree: str
    specialisations: dict[str, list[str]]

    def __init__(
        self,
        name: str,
        course_dates: dict[tuple[int, int], list[str]],
        course_reqs: dict[str, list[str]],
        courses: list[str],
        degree: str,
        specialisations: dict[str, list[str]],
    ):
        self.name = name
        self.course_dates = course_dates
        self.course_reqs = course_reqs
        self.courses = courses
        self.degree = degree
        self.specialisations = specialisations

File: frontend/types/plan.ts
export interface Plan {
  name: string,
  id: string,
  degree: Degree,

  percentage: number,

  start_sem: number
  end_year: number,

  courses: PlannedCourses,

  specialisations: object,
}

export interface PlannedCourses {
  [year: string]: {
    sem: CourseData[];
  }
};

export interface CourseData {
  code: string,
  pos: number,
  part_id: string
}

export interface Degree {
  degree_id: string,
  degree_code: number,
  year: number,
  title: string,
  degree_url: string,
}
